packages/wdio-devtools-service/src/index.js

```diff
@@ -29,23 +36,133 @@ export default class DevToolsService {
+            this.commandHandler = new CommandHandler(this.client, global.browser)
+            this.devtoolsDriver = await DevToolsDriver.attach(`http://${debuggerAddress.host}:${debuggerAddress.port}`)
+            this.traceGatherer = new TraceGatherer(this.devtoolsDriver)
+
+            const session = await this.devtoolsDriver.getCDPSession()
+            session.on('Page.loadEventFired', ::this.traceGatherer.onLoadEventFired)
+            session.on('Page.frameNavigated', ::this.traceGatherer.onFrameNavigated)
+
+            const page = await this.devtoolsDriver.getActivePage()
+            page.on('requestfailed', ::this.traceGatherer.onFrameLoadFail)
+
+            /**
+             * enable domains for client
+             */
+            await Promise.all(['Page', 'Network', 'Console'].map(
+                (domain) => Promise.all([
+                    session.send(`${domain}.enable`),
+                    this.client[domain]['enable']()
+                ])
+            ))
+
+            this.devtoolsGatherer = new DevtoolsGatherer()
+            this.client.on('event', ::this.devtoolsGatherer.onMessage)
+
+            log.info(`Connected to Chrome on ${debuggerAddress.host}:${debuggerAddress.port}`)
         } catch (err) {
             log.error(`Couldn't connect to chrome: ${err.stack}`)
             return
         }
 
+        global.browser.addCommand('enablePerformanceAudits', ::this._enablePerformanceAudits)
+        global.browser.addCommand('disablePerformanceAudits', ::this._disablePerformanceAudits)
+
         /**
-         * enable network and page domain for resource analysis
+         * allow user to work with Puppeteer directly
          */
-        await this.commandHandler.cdp('Network', 'enable')
-        await this.commandHandler.cdp('Page', 'enable')
+        global.browser.addCommand('getPuppeteer',
+            () => this.devtoolsDriver)
+    }
+
+    async beforeCommand (commandName, params) {
+        if (!this.shouldRunPerformanceAudits || !this.traceGatherer || this.traceGatherer.isTracing || !TRACE_COMMANDS.includes(commandName)) {
+            return
+        }
+
+        /**
+         * set browser profile
+         */
+        this._setThrottlingProfile(this.networkThrottling, this.cpuThrottling, this.cacheEnabled)
+
+        const url = commandName === 'navigateTo' ? params[0] : 'click transition'
+        return this.traceGatherer.startTracing(url)
+    }
+
+    async afterCommand (commandName) {
+        if (!this.traceGatherer || !this.traceGatherer.isTracing || !TRACE_COMMANDS.includes(commandName)) {
+            return
+        }
+
+        /**
+         * update custom commands once tracing finishes
+         */
+        this.traceGatherer.once('tracingComplete', (traceEvents) => {
+            const auditor = new Auditor(traceEvents, this.devtoolsGatherer.getLogs())
+            auditor.updateCommands(global.browser)
+        })
+
+        return new Promise((resolve) => {
```

Comment:
I think using `Promise` directly it is not necessary here

Reply:
it is, we need to defer the continuation until the tracing has been finished .. this can only be done by returning a promise